NE-978: ingress: Add ReloadInterval Option (#1184)
openshift/enhancements: enhancements/ingress/haproxy-reload-interval.md

---
title: haproxy-reload-interval
authors:
  - "@Ethany-RH"

> reviewers:

Reviewer: These can be the same as other haproxy enhancement proposals:
```
reviewers:
  - "@frobware"
  - "@knobunc"  
  - "@Miciah"
  - "@candita"
  - "@miheer"
  - "@rfredette"
  - "@deads2k"
approvers:
  - "@frobware"
  - "@Miciah"
  - "@candita"
```

  - "?"
approvers:
  - "?"

> api-approvers: # necessary?

Reviewer: - "@deads2k"

  - "?"
creation-date: 2022-07-01

> last-updated: 2022-07-01

Reviewer: Update this.

tracking-link:
  - "https://issues.redhat.com/browse/NE-586"
see-also:
replaces:
superseded-by:
---

# Reload Interval in HAProxy

## Release Signoff Checklist

- [ ] Enhancement is `implementable`.
- [ ] Design details are appropriately documented from clear requirements.
- [ ] Test plan is defined.
- [ ] graduation criteria for dev preview, tech preview, GA
- [ ] User-Facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/).

## Summary

> Add an API field to configure OpenShift router's `RELOAD_INTERVAL` environment variable so that administrators can define the minimum frequency the router is allowed to reload to accept new changes.

Reviewer: ```suggestion
Add an API field to configure OpenShift router's `RELOAD_INTERVAL` environment variable so that administrators can define the minimum frequency the router is allowed to reload in order to publish new changes.
```

> OpenShift router currently hard-codes this reload interval to 5s. It should be possible for administrators to tune this value as necessary. Based on the processes run in the cluster and the frequency that it sees new changes, increasing the minimum interval at which the router is allowed to reload can improve its efficiency.

Reviewer: ```suggestion
OpenShift router currently hard-codes this reload interval to 5 seconds. It should be possible for administrators to tune this value as necessary. Based on the processes run in the cluster and the frequency that it sees new changes, increasing the minimum interval at which the router is allowed to reload can improve its efficiency.
```

This proposal extends the existing IngressController API to add a tuning option for the reload interval.

## Motivation

> When there is an update to a route or endpoint in the cluster, the configuration for HAProxy changes, requiring that it reload for those changes to take effect. When HAProxy reloads to generate a new process with the updated configuration, it must keep the old process running until all its connections die. As such, frequent reloading increases the rate of accumulation of HAProxy processes, particularly if it has to handle many long-lived connections. The default reload interval is set as 5 seconds, which is too low for some scenarios, so it is important that administrators can extend this time interval.

Reviewer: ```suggestion
When there is an update to a route or endpoint in the cluster, the configuration for HAProxy changes, requiring that it reload for those changes to take effect. When HAProxy reloads to generate a new process with the updated configuration, it must keep the old process running until all its connections die. As such, too-frequent reloading can increase the rate of accumulation of HAProxy processes, particularly if it has to handle many long-lived connections. The default reload interval is set as 5 seconds, which is too low for some scenarios, so it is important that administrators can extend this time interval.
```

> In addition, HAProxy's roundrobin balancing starts over from the first server every time HAProxy reloads. Thus, another motivating factor is that frequent reloads can cause load imbalance on a backend's servers when using the roundrobin balancing algorithm.

Reviewer 2: This affects other balancing algorithms, not just roundrobin.  However, this is a good example—it's easy to explain how the issue affects roundrobin.

> In addition, HAProxy's roundrobin balancing starts over from the first server every time HAProxy reloads. Thus, another motivating factor is that frequent reloads can cause load imbalance on a backend's servers when using the roundrobin balancing algorithm.

Reviewer: ```suggestion
In addition, HAProxy's roundrobin load balancing starts over from the first server every time HAProxy reloads. Thus, another motivating factor is that too-frequent reloads can cause load imbalance on a backend's servers when using the roundrobin balancing algorithm.
```

### Goals

1. Enable the configuration of a reload interval via the `IngressControllerSpec`, specifically via the `IngressControllerSpecTuningOptions`, with a new parameter `ReloadInterval`. `ReloadInterval` exposes OpenShift router's internal environment variable `RELOAD_INTERVAL` as an API that the cluster administrator can set.

> 2. Leave the default interval at 5 seconds so that we do not perturb the behavior of existing clusters, particularly during upgrades.

Reviewer: Add a goal that explains the limits you imposed.

// of ingress controller pods
type IngressControllerTuningOptions struct {
  	...

	// reloadInterval defines the minimum interval at which the router is allowed to reload
	// to accept new changes. Increasing this value can prevent the accumulation of
	// HAProxy processes, depending on the scenario. Increasing this interval can
	// also lessen load imbalance on a backend's servers when using the roundrobin
	// balancing algorithm. Alternatively, decreasing this value may decrease latency
	// since updates to HAProxy's configuration can take effect more quickly.
	//
	// The value must be a time duration value; see <https://pkg.go.dev/time#ParseDuration>.
	// Currently, the minimum value allowed is 1s, and the maximum allowed value is
	// 120s. Minimum and maximum allowed values may change in future versions of OpenShift.
	// Note that if a duration outside of these bounds is provided, the value of reloadInterval
	// will be capped/floored and not rejected (e.g. a duration of over 120s will be capped to
	// 120s; the IngressController will not reject and replace this disallowed value with
	// the default).
	//

> 	// A nil or zero value for reloadInterval tells the IngressController to choose the default,

Reviewer: Make sure this section is up-to-date with what merged into openshift/api.

	// which is currently 5s and subject to change.
	//
	// This field expects an unsigned duration string of decimal numbers, each with optional
	// fraction and a unit suffix, e.g. "100s", "1m30s". Valid time units are "s" and "m".
	//
	// Note: Setting a value significantly larger than the default of 5s can cause latency
	// in observing updates to routes and their endpoints. HAProxy's configuration will
	// be reloaded less frequently, and newly created routes will not be served until the
	// subsequent reload.
	//
	// +kubebuilder:validation:Optional
	// +kubebuilder:validation:Pattern=^(0|([0-9]+(\.[0-9]+)?(s|m))+)$
	// +kubebuilder:validation:Type:=string
	// +optional
	ReloadInterval metav1.Duration `json:"reloadInterval,omitempty"`
}
```

### Implementation Details / Notes / Constraints

To expose the `ReloadInterval` in HAProxy, the environment variable `RELOAD_INTERVAL` will be added to the environment in [desiredRouterDeployment](https://github.com/openshift/cluster-ingress-operator/blob/master/pkg/operator/controller/ingress/deployment.go).

> The HAProxy template will not be modified.

Reviewer: Mention some of the details from the godoc here, specific min/max/default values and parsing requirements.

Reviewer: Also mention the unsupported config override that has been removed.  We don't support if someone has an issue, but it doesn't mean we shouldn't at least mention it here.

### Risks and Mitigation

> A risk in this proposal is that customers who set a long reload interval to decrease the potential memory usage of HAProxy instances may inadverdently create latency issues in the cluster. Setting a large value for the reload interval can cause significant latency in observing updates to routes and their endpoints. This is because HAProxy's configuration will be reloaded less frequently, and newly created routes will not be served until the subsequent reload.

Reviewer 2: ```suggestion
A risk in this proposal is that customers who set a long reload interval to decrease the potential memory usage of HAProxy instances may inadvertently create latency issues in the cluster. Setting a large value for the reload interval can cause significant latency in observing updates to routes and their endpoints. This is because HAProxy's configuration will be reloaded less frequently, and newly created routes will not be served until the subsequent reload.
```
The linter will want you to wrap this line.

To mitigate this risk, we will set a lower cap (than other interval environment variables) to limit the largest time interval that `reloadInterval` will accept. In addition, we have also included a note in the API godoc warning users of the possible risk in setting a large reload interval.

### Drawbacks

## Design Details

### Test Plan

Unit testing can validate that `desiredRouterDeployment` sets the `RELOAD_INTERVAL` environment variable correctly. Unit testing can also validate that `capReloadIntervalValue` in deployment.go sets minimum and maximum caps correctly.

E2E Tests

> 1. Create a new IngressController. Wait for an ingress controller pod to be deployed.

Reviewer: You can paraphrase what is tested end to end here.

### Graduation Criteria

#### Dev Preview -> Tech Preview

#### Tech Preview -> GA

#### Removing a deprecated feature

> ### Upgrade/Downgrade Strategy

Reviewer: You should mention the unsupported config overrides here too.  If they used unsupported config overrides and upgrade without setting the `reloadInterval`, they will lose their values.  This is fine, but worth mentioning in here.

Upgrading from a previous release that does not have `Spec.TuningOptions.ReloadInterval` will leave the field empty, which is an acceptable state. With the field empty, the default value of 5s will be used.

If `Spec.TuningOptions.ReloadInterval` is set when downgrading to a release without the field, the value will be discarded, and the ingress controller will revert to the previous default of 5s.

### Version Skew Strategy

N/A

### Operations Aspects of API Extensions

#### Failure Modes

N/A

#### Support Procedures

> If the frequency of reloads compromises the performance of HAProxy, and the revert to default values does not fix it, then that is indicative of another issue.

Reviewer: ```suggestion
If the frequency of reloads compromises the performance of HAProxy, and the revert to the default value of 5 seconds does not fix the performance problem, then that is indicative of another issue.
```

    (*appsv1.Deployment, error){
        ...
        reloadInterval := 5 * time.Second
	    if ci.Spec.TuningOptions.ReloadInterval != nil && ci.Spec.TuningOptions.ReloadInterval.Duration >= 1 * time.Second && ci.Spec.TuningOptions.ReloadInterval.Duration <= 720 * time.Hour {
		    reloadInterval = ci.Spec.TuningOptions.ReloadInterval.Duration
	    }

> 	    env = append(env, corev1.EnvVar{Name: RouterReloadIntervalEnvName, Value: strconv.Itoa(int(reloadInterval.Seconds()))})

Reviewer 2: Does this work?
```suggestion
	    env = append(env, corev1.EnvVar{Name: RouterReloadIntervalEnvName, Value: fmt.Sprintf("%v", reloadInterval.Seconds())})
```
That avoids truncating to an integer and formats as a decimal (e.g., "10500ms" gets formatted as "10.5" seconds), and I think `RELOAD_INTERVAL` will properly handle values like "10.5", based on <https://go.dev/play/p/4w625xB8suO>.

Author: In that case, I can just take out the `int()` and just have `strconv.Itoa(reloadInterval.Seconds())`. Should I still use `fmt.Sprintf("%v", reloadInterval.Seconds())` instead?

Author: Nevermind, `strconv.Itoa()` doesn't work on floats. I'll change my code to use `fmt.Sprintf("%v", reloadInterval.Seconds())`, thanks!

### Implementation Details / Notes / Constraints

> To expose the `ReloadInterval` in HAProxy, the environment variable `RELOAD_INTERVAL` will be added to the environment in [desiredRouterDeployment](https://github.com/openshift/cluster-ingress-operator/blob/master/pkg/operator
> controller/ingress/deployment.go). desiredRouterDeployment will also ensure that the value passed to `RELOAD_INTERVAL` is within the range of 1 second - 120 seconds. If `ReloadInterval` is not set by the user or has a zero value,

Reviewer 2: The link got broken during wrapping.  
```suggestion
To expose the `ReloadInterval` in HAProxy, the environment variable `RELOAD_INTERVAL` will be added to the environment in
[desiredRouterDeployment](https://github.com/openshift/cluster-ingress-operator/blob/master/pkg/operator/controller/ingress/deployment.go).
`desiredRouterDeployment` will also ensure that the value passed to `RELOAD_INTERVAL` is within the range of 1 second - 120 seconds. If `ReloadInterval` is not set by the user or has a zero value,
```

the default of 5 seconds will be used.
The `reloadInterval` field expects an unsigned duration string of decimal numbers, each with optional fraction and a unit suffix, e.g. "100s", "1m30s". Valid time units are "s" and "m". This will be enforced by the API's validation.
The HAProxy template will not be modified.
Since `ReloadInterval` will be exposed as a tuning option in `IngressControllerTuningOptions`, it will no longer be an unsupported config and will be removed from `unsupportedConfigOverrides` in desiredRouterDeployment.

### Risks and Mitigations

> A risk in this proposal is that customers who set a long reload interval to decrease the potential memory usage of HAProxy instances may inadverdently

Reviewer 2: ```suggestion
A risk in this proposal is that customers who set a long reload interval to decrease the potential memory usage of HAProxy instances may inadvertently
```
